### DISCUSSION TITLE
Stuck with transaction error when creating multiple records

### DISCUSSION THREAD

**Discussion by KiriLucas** [prisma/prisma#11277] [Q&A] [answered]
Hi everyone!

I need some light on $transactions API. I just started using Prisma and still kinda confused (but amazed).
I have 5 recently created tables: `header`, `position`, `sequence`, `operatio`n and `component`. My applicattion receives it from an external application all at once, since it's a huge entity called "Production Order" and each table is a part of it and I need to save it with the following relations:

1. `Position`, `sequence ` and `component` tables are all dependants on the id field of the header table (fk)
2. There might be ONLY one `header` per Production Order, but there may be multiple positions, sequences and components linked to the header.
3. `Operations` is dependant on `sequence` only and it might also have multiples operations for a single sequence
4. Header has an id field, which must be an fk to every other table (except the operations ones)

Unfortunately, I've been get the following error when trying to create a new record:
```typescript
node_1      | /app/node_modules/@prisma/client/runtime/index.js:39601
node_1      |           throw new Error(`All elements of the array need to be Prisma Client promises. Hint: Please make sure you are not awaiting the Prisma client calls you intended to pass in the $transaction 
function.`);
node_1      |                 ^
node_1      | Error: All elements of the array need to be Prisma Client promises. Hint: Please make sure you are not awaiting the Prisma client calls you intended to pass in the $transaction function.
```

Now's the weird part, this is how I'm trying to save it:

```typescript
    await async.mapSeries(
      Object.entries(productionOrderStructure),
      async ([key, { inject, table, method }]) => {
        const injections = (inject || []).reduce((red, { getSaved, as }) => {
          red[as] = get(savedData, getSaved);
          return red;
        }, {});

        const transaction = await this.repository.transaction(
          productionOrderObject[key].map((data) =>
            this.repository[method]({
              data: {
                ...data,
                ...injections,
              },
            }),
          ),
        );

        savedData[key] = transaction;

        return transaction;
      },
    );
```
Pay close attention to that `const transaction` part, because `this.repository.transaction`  and ``this.repository[method]`  are methods from my repository class:
```typescript
   // [...]
  constructor(private readonly prisma: PrismaService) {}
   // [...]

  async transaction(array){
    return await this.prisma.transaction(array)
  }

  async saveHeader(data) { //this.repository.saveHeader - [method], in this iteration, is saveHeader
    return this.prisma.productionOrder.create(data);
  }
```

And `this.prisma` is my PrismaService class, which looks like this:

```typescript
export class PrismaService extends PrismaClient implements OnModuleInit{

  async onModuleInit() {
    await this.$connect();
  }

  async enableShutdownHooks(app: INestApplication) {
    this.$on('beforeExit', async () => {
      await app.close();
    });
  }

    async saveHeader(data) {
      return this.productionOrder.create(data);
    }

    async transaction(array){
      return await this.$transaction(array)
    }
```

But if instead of going trough my repository, I directly call my PrismaService within that transaction:
```typescript
this.prismaService[method].create({ // [...]
```
It instantly works, now, I'm not sure if I'm being silly or if this might be a bug, but I would certainly appreciate some enlightenment as I was the one to bring Prisma to the project I work at (and it's doing amazing so far!)

**Comment by austincrim** [comment, marked as answer]
Hey @KiriLucas 👋 ,

I think the root issue here is that you are actually passing an array of regular JavaScript promises that **wrap** Prisma promises. Since your `saveHeader` function is `async`, it's going to return a **JavaScript** promise that wraps a Prisma promise. The `transaction` method is expecting an array of **Prisma** promises.

In your current code, the resulting array that gets passed to `this.repository.transaction` looks like this:
```ts
this.repository.transaction([Promise<PrismaPromise>, Promise<PrismaPromise>...])
```
It needs to look like this instead:
```ts
this.repository.transaction([PrismaPromise, PrismaPromise...])
```

If you remove the `async` keyword from your repository methods like `saveHeader`, it should directly return a Prisma promise and everything should work.

Please let me know if I can clarify anything further!

**Reply by KiriLucas** [reply]
Hmmm... Interesting, it seems like that was the issue indeed. Thanks for your help!

**Reply by Fateh-bak** [reply]
thank you for the solution!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
